File: server/python/cluster_setup.py
from __future__ import annotations

import asyncio
import base64
import hashlib
import os
from asyncio.events import AbstractEventLoop
from collections.abc import Mapping
from typing import Any

import rpc
import rpc_reader
from typing import TypedDict, Callable


class ClusterObject(TypedDict):
    id: str
    address: str
    port: int
    proxyId: str
    sourceKey: str
    sha256: str


def isClusterAddress(address: str):
    return not address or address == os.environ.get("SCRYPTED_CLUSTER_ADDRESS", None)


def getClusterPeerKey(address: str, port: int):
    return f"{address}:{port}"


class ClusterSetup:
    def __init__(self, loop: AbstractEventLoop, peer: rpc.RpcPeer):
        self.loop = loop
        self.peer = peer
        self.clusterId: str = None
        self.clusterSecret: str = None
        self.clusterAddress: str = None
        self.clusterPort: int = None
        self.SCRYPTED_CLUSTER_ADDRESS: str = None
        self.clusterPeers: Mapping[str, asyncio.Future[rpc.RpcPeer]] = {}

    async def resolveObject(self, id: str, sourceKey: str):
        sourcePeer: rpc.RpcPeer = (
            self.peer
            if not sourceKey
            else await rpc.maybe_await(self.clusterPeers.get(sourceKey, None))
        )
        if not sourcePeer:
            return
        return sourcePeer.localProxyMap.get(id, None)

    async def connectClusterObject(self, o: ClusterObject):
        sha256 = self.computeClusterObjectHash(o)
        if sha256 != o["sha256"]:
            raise Exception("secret incorrect")
        return await self.resolveObject(
            o.get("proxyId", None), o.get("sourceKey", None)
        )

    def onProxySerialization(
        self, peer: rpc.RpcPeer, value: Any, sourceKey: str = None
    ):
        properties: dict = rpc.RpcPeer.prepareProxyProperties(value) or {}
        clusterEntry = properties.get("__cluster", None)
        proxyId: str
        existing = peer.localProxied.get(value, None)
        if existing:
            proxyId = existing["id"]
        else:
            proxyId = (
                clusterEntry and clusterEntry.get("proxyId", None)
            ) or rpc.RpcPeer.generateId()

        if clusterEntry:
            if (
                isClusterAddress(clusterEntry.get("address", None))
                and self.clusterPort == clusterEntry["port"]
                and sourceKey != clusterEntry.get("sourceKey", None)
            ):
                clusterEntry = None

        if not clusterEntry:
            clusterEntry: ClusterObject = {
                "id": self.clusterId,
                "proxyId": proxyId,
                "address": self.SCRYPTED_CLUSTER_ADDRESS,
                "port": self.clusterPort,
                "sourceKey": sourceKey,
            }
            clusterEntry["sha256"] = self.computeClusterObjectHash(clusterEntry)
            properties["__cluster"] = clusterEntry

        return proxyId, properties

    async def initializeCluster(self, options: dict):
        if self.clusterPort:
            return
        self.clusterId = options["clusterId"]
        self.clusterSecret = options["clusterSecret"]
        self.clusterWorkerId = options.get("clusterWorkerId", None)
        self.SCRYPTED_CLUSTER_ADDRESS = os.environ.get("SCRYPTED_CLUSTER_ADDRESS", None)

        async def handleClusterClient(
            reader: asyncio.StreamReader, writer: asyncio.StreamWriter
        ):
            clusterPeerAddress, clusterPeerPort = writer.get_extra_info("peername")
            clusterPeerKey = getClusterPeerKey(clusterPeerAddress, clusterPeerPort)
            rpcTransport = rpc_reader.RpcStreamTransport(reader, writer)
            peer: rpc.RpcPeer
            peer, peerReadLoop = await rpc_reader.prepare_peer_readloop(
                self.loop, rpcTransport
            )
            # set all params from self.peer
            for key, value in self.peer.params.items():
                peer.params[key] = value
            peer.onProxySerialization = lambda value: self.onProxySerialization(
                peer, value, clusterPeerKey
            )
            future: asyncio.Future[rpc.RpcPeer] = asyncio.Future()
            future.set_result(peer)
            self.clusterPeers[clusterPeerKey] = future
            peer.params["connectRPCObject"] = lambda o: self.connectClusterObject(o)
            try:
                await peerReadLoop()
            except:
                pass
            finally:
                self.clusterPeers.pop(clusterPeerKey)
                peer.kill("cluster client killed")
                writer.close()

        clusterRpcServerInfo = await cluster_listen_zero(handleClusterClient)
        self.clusterPort = clusterRpcServerInfo["port"]
        self.peer.onProxySerialization = lambda value: self.onProxySerialization(
            self.peer, value, None
        )
        del self.peer.params["initializeCluster"]

    def computeClusterObjectHash(self, o: ClusterObject) -> str:
        m = hashlib.sha256()
        m.update(
            bytes(
                # The use of ` o.get(key, None) or '' ` is to ensure that optional fields
                # are omitted from the hash, matching the JS implementation. Otherwise, since
                # the dict may contain the keys initialized to None, ` o.get(key, '') ` would
                # return None instead of ''.
                f"{o['id']}{o.get('address', None) or ''}{o['port']}{o.get('sourceKey', None) or ''}{o['proxyId']}{self.clusterSecret}",
                "utf8",
            )
        )
        return base64.b64encode(m.digest()).decode("utf-8")

    def ensureClusterPeer(self, address: str, port: int):
        if isClusterAddress(address):
            address = "127.0.0.1"
        clusterPeerKey = getClusterPeerKey(address, port)
        clusterPeerPromise = self.clusterPeers.get(clusterPeerKey)
        if clusterPeerPromise:
            return clusterPeerPromise

        async def connectClusterPeer():
            try:
                reader, writer = await asyncio.open_connection(address, port)
                sourceAddress, sourcePort = writer.get_extra_info("sockname")
                if (
                    sourceAddress != self.SCRYPTED_CLUSTER_ADDRESS
                    and sourceAddress != "127.0.0.1"
                ):
                    print("source address mismatch", sourceAddress)
                rpcTransport = rpc_reader.RpcStreamTransport(reader, writer)
                clusterPeer, peerReadLoop = await rpc_reader.prepare_peer_readloop(
                    self.loop, rpcTransport
                )
                # set all params from self.peer
                for key, value in self.peer.params.items():
                    clusterPeer.params[key] = value
                clusterPeer.onProxySerialization = (
                    lambda value: self.onProxySerialization(
                        clusterPeer, value, clusterPeerKey
                    )
                )
            except:
                self.clusterPeers.pop(clusterPeerKey)
                raise

            async def run_loop():
                try:
                    await peerReadLoop()
                except:
                    pass
                finally:
                    self.clusterPeers.pop(clusterPeerKey)

            asyncio.run_coroutine_threadsafe(run_loop(), self.loop)
            return clusterPeer

        clusterPeerPromise = self.loop.create_task(connectClusterPeer())

        self.clusterPeers[clusterPeerKey] = clusterPeerPromise
        return clusterPeerPromise

    async def connectRPCObject(self, value):
        __cluster = getattr(value, "__cluster")
        if type(__cluster) is not dict:
            return value

        clusterObject: ClusterObject = __cluster

        if clusterObject.get("id", None) != self.clusterId:
            return value

        address = clusterObject.get("address", None)
        port = clusterObject["port"]
        proxyId = clusterObject["proxyId"]
        if port == self.clusterPort:
            return await self.connectRPCObject(clusterObject)

        clusterPeerPromise = self.ensureClusterPeer(address, port)

        try:
            clusterPeer = await clusterPeerPromise
            weakref = clusterPeer.remoteWeakProxies.get(proxyId, None)
            existing = weakref() if weakref else None
            if existing:
                return existing

            peerConnectRPCObject = clusterPeer.tags.get("connectRPCObject")
            if not peerConnectRPCObject:
                peerConnectRPCObject = await clusterPeer.getParam("connectRPCObject")
                clusterPeer.tags["connectRPCObject"] = peerConnectRPCObject
            newValue = await peerConnectRPCObject(clusterObject)
            if not newValue:
                raise Exception("rpc object not found?")
            return newValue
        except Exception as e:
            return value


class ClusterServerListener(TypedDict):
    server: asyncio.Server
    port: int


async def cluster_listen_zero(
    callback: Callable[[asyncio.StreamReader, asyncio.StreamWriter]]
) -> ClusterServerListener:
    SCRYPTED_CLUSTER_ADDRESS = os.getenv("SCRYPTED_CLUSTER_ADDRESS")
    if not SCRYPTED_CLUSTER_ADDRESS:
        server = await asyncio.start_server(callback, host=None, port=0)
        port = server.sockets[0].getsockname()[1]
        return {
            "server": server,
            "port": port,
        }

    # need to listen on the cluster address and 127.0.0.1 on the same port.
    retries = 5
    while retries > 0:
        cluster_server = await asyncio.start_server(
            callback, host=SCRYPTED_CLUSTER_ADDRESS, port=0
        )
        port = cluster_server.sockets[0].getsockname()[1]

        try:
            print('trying to bind to port', port)
            local_server = await asyncio.start_server(
                callback, host="127.0.0.1", port=port
            )

            future = asyncio.ensure_future(local_server.wait_closed())
            future.add_done_callback(lambda: local_server.close())

            return {
                "server": cluster_server,
                "port": port,
            }
        except:
            # Port may be in use, keep trying.
            cluster_server.close()
            retries -= 1

    raise Exception("failed to bind to cluster address.")

File: server/src/cluster/cluster-setup.ts
import { once } from 'events';
import net from 'net';
import os from 'os';
import worker_threads from 'worker_threads';
import { Deferred } from '../deferred';
import { listenZero } from '../listen-zero';
import { NodeThreadWorker } from '../plugin/runtime/node-thread-worker';
import { RpcPeer } from "../rpc";
import { createDuplexRpcPeer } from "../rpc-serializer";
import { computeClusterObjectHash } from "./cluster-hash";
import { ClusterObject, ConnectRPCObject } from "./connect-rpc-object";

function getClusterPeerKey(address: string, port: number) {
    return `${address}:${port}`;
}

export function isClusterAddress(address: string) {
    return !address || address === process.env.SCRYPTED_CLUSTER_ADDRESS;
}

export function getClusterObject(clusterId: string, value: any) {
    const clusterObject: ClusterObject = value?.__cluster;
    if (clusterObject?.id !== clusterId)
        return;
    return clusterObject;
}

async function peerConnectRPCObject(peer: RpcPeer, o: ClusterObject) {
    let peerConnectRPCObject: Promise<ConnectRPCObject> = peer.tags['connectRPCObject'];
    if (!peerConnectRPCObject) {
        peerConnectRPCObject = peer.getParam('connectRPCObject');
        peer.tags['connectRPCObject'] = peerConnectRPCObject;
    }
    const resolved = await peerConnectRPCObject;
    return resolved(o);
}

export function setupCluster(peer: RpcPeer) {
    const SCRYPTED_CLUSTER_ADDRESS = process.env.SCRYPTED_CLUSTER_ADDRESS;
    let clusterId: string;
    let clusterSecret: string;
    let clusterPort: number;
    let clusterWorkerId: string;

    // all cluster clients, incoming and outgoing, connect with random ports which can be used as peer ids
    // on the cluster server that is listening on the actual port.
    // incoming connections: use the remote random/unique port
    // outgoing connections: use the local random/unique port
    const clusterPeers = new Map<string, Promise<RpcPeer>>();

    peer.killedSafe.finally(() => {
        for (const cp of clusterPeers.values()) {
            cp.then(c => c.kill()).catch(() => { });
        }
    });

    const resolveObject = async (id: string, sourceKey: string) => {
        const sourcePeer = sourceKey
            ? await clusterPeers.get(sourceKey)
            : peer;
        if (!sourcePeer)
            console.error('source peer not found', sourceKey);
        const ret = sourcePeer?.localProxyMap.get(id);
        if (!ret) {
            console.error('source key not found', sourceKey, id);
            return;
        }
        return ret;
    }

    const connectClusterObject = async (o: ClusterObject) => {
        const sha256 = computeClusterObjectHash(o, clusterSecret);
        if (sha256 !== o.sha256)
            throw new Error('secret incorrect');
        return resolveObject(o.proxyId, o.sourceKey);
    }

    const ensureClusterPeer = (address: string, connectPort: number) => {
        if (isClusterAddress(address))
            address = '127.0.0.1';

        const clusterPeerKey = getClusterPeerKey(address, connectPort);
        let clusterPeerPromise = clusterPeers.get(clusterPeerKey);
        if (clusterPeerPromise)
            return clusterPeerPromise;

        clusterPeerPromise = (async () => {
            const socket = net.connect(connectPort, address);
            socket.on('close', () => clusterPeers.delete(clusterPeerKey));

            try {
                await once(socket, 'connect');
                const { address: sourceAddress } = (socket.address() as net.AddressInfo);
                if (sourceAddress !== SCRYPTED_CLUSTER_ADDRESS && sourceAddress !== '127.0.0.1') {
                    // source address may end with .0 if its a gateway into docker.
                    if (!sourceAddress.endsWith('.0'))
                        console.warn("source address mismatch", sourceAddress);
                }

                const clusterPeer = createDuplexRpcPeer(peer.selfName, clusterPeerKey, socket, socket);
                // set params from the primary peer, needed for get getRemote in cluster mode.
                Object.assign(clusterPeer.params, peer.params);
                clusterPeer.onProxySerialization = (value) => onProxySerialization(clusterPeer, value, clusterPeerKey);
                return clusterPeer;
            }
            catch (e) {
                console.error('failure ipc connect', e);
                socket.destroy();
                throw e;
            }
        })();

        clusterPeers.set(clusterPeerKey, clusterPeerPromise);
        return clusterPeerPromise;
    };


    const tidChannels = new Map<number, Deferred<worker_threads.MessagePort>>();
    const tidPeers = new Map<number, Promise<RpcPeer>>();

    peer.killedSafe.finally(() => {
        for (const cp of tidPeers.values()) {
            cp.then(c => c.kill()).catch(() => { });
        }
    });

    function connectTidPeer(tid: number) {
        let peerPromise = tidPeers.get(tid);
        if (peerPromise)
            return peerPromise;
        let tidDeferred = tidChannels.get(tid);
        // if the tid port is not available yet, request it.
        if (!tidDeferred) {
            tidDeferred = new Deferred<worker_threads.MessagePort>();
            tidChannels.set(tid, tidDeferred);

            if (mainThreadPort) {
                // request the connection via the main thread
                mainThreadPort.postMessage({
                    threadId: tid,
                });
            }
        }

        const threadPeerKey = `thread:${tid}`;
        function peerCleanup() {
            clusterPeers.delete(threadPeerKey);
        }
        peerPromise = tidDeferred.promise.then(port => {
            const threadPeer = NodeThreadWorker.createRpcPeer(peer.selfName, threadPeerKey, port);
            // set params from the primary peer, needed for get getRemote in cluster mode.
            Object.assign(threadPeer.params, peer.params);
            threadPeer.onProxySerialization = value => onProxySerialization(threadPeer, value, threadPeerKey);

            threadPeer.params.connectRPCObject = connectClusterObject;

            function cleanup(message: string) {
                peerCleanup();
                tidChannels.delete(tid);
                tidPeers.delete(tid);
                threadPeer.kill(message);
            }
            port.on('close', () => cleanup('connection closed.'));
            port.on('messageerror', () => cleanup('message error.'));
            return threadPeer;
        });
        peerPromise.catch(() => peerCleanup());
        clusterPeers.set(threadPeerKey, peerPromise);
        tidPeers.set(tid, peerPromise);

        return peerPromise;
    }

    const mainThreadPort: worker_threads.MessagePort = worker_threads.isMainThread ? undefined : worker_threads.workerData.mainThreadPort;
    if (!worker_threads.isMainThread) {
        // the main thread port will send messages with a thread port when a thread wants to initiate a connection.
        mainThreadPort.on('message', async (message: { port: worker_threads.MessagePort, threadId: number }) => {
            const { port, threadId } = message;
            let tidDeferred = tidChannels.get(threadId);
            if (!tidDeferred) {
                tidDeferred = new Deferred<worker_threads.MessagePort>();
                tidChannels.set(threadId, tidDeferred);
            }
            tidDeferred.resolve(port);
            connectTidPeer(threadId);
        });
    }

    async function connectIPCObject(clusterObject: ClusterObject, tid: number) {
        // if the main thread is trying to connect to an object,
        // the argument order matters here, as the connection attempt looks at the
        // connectThreadId to see if the target is main thread.
        if (worker_threads.isMainThread)
            mainThreadBrokerConnect(tid, worker_threads.threadId);
        const clusterPeer = await connectTidPeer(tid);
        const existing = clusterPeer.remoteWeakProxies[clusterObject.proxyId]?.deref();
        if (existing)
            return existing;
        return peerConnectRPCObject(clusterPeer, clusterObject);
    }

    const brokeredConnections = new Set<string>();
    const workers = new Map<number, worker_threads.MessagePort>();
    function mainThreadBrokerConnect(threadId: number, connectThreadId: number) {
        if (worker_threads.isMainThread && threadId === worker_threads.threadId) {
            const msg = 'invalid ipc, main thread cannot connect to itself';
            console.error(msg);
            throw new Error(msg);
        }
        // both workers nay initiate connection to each other at same time, so this
        // is a synchronization point.
        const key = JSON.stringify([threadId, connectThreadId].sort());
        if (brokeredConnections.has(key))
            return;

        brokeredConnections.add(key);

        const worker = workers.get(threadId);
        const connect = workers.get(connectThreadId);
        const channel = new worker_threads.MessageChannel();

        worker.postMessage({
            port: channel.port1,
            threadId: connectThreadId,
        }, [channel.port1]);

        if (connect) {
            connect.postMessage({
                port: channel.port2,
                threadId,
            }, [channel.port2]);
        }
        else if (connectThreadId === worker_threads.threadId) {
            connectTidPeer(threadId);
            const deferred = tidChannels.get(threadId);
            deferred.resolve(channel.port2);
        }
        else {
            channel.port2.close();
        }
    }

    function mainThreadBrokerRegister(workerPort: worker_threads.MessagePort, threadId: number) {
        workers.set(threadId, workerPort);

        // this is main thread, so there will be two types of requests from the child: registration requests from grandchildren and connection requests.
        workerPort.on('message', async (message: { port: worker_threads.MessagePort, threadId: number }) => {
            const { port, threadId: connectThreadId } = message;

            if (port) {
                mainThreadBrokerRegister(port, connectThreadId);
            }
            else {
                mainThreadBrokerConnect(threadId, connectThreadId);
            }
        });
    }

    const connectRPCObject = async (value: any) => {
        const clusterObject = getClusterObject(clusterId, value);
        if (!clusterObject)
            return value;
        const { address, port, proxyId } = clusterObject;
        // handle the case when trying to connect to an object is on this cluster node,
        // returning the actual object, rather than initiating a loopback connection.
        if (port === clusterPort)
            return connectClusterObject(clusterObject);

        // can use worker to worker ipc if the address and pid matches and both side are node.
        if (address === SCRYPTED_CLUSTER_ADDRESS && proxyId.startsWith('n-')) {
            const parts = proxyId.split('-');
            const pid = parseInt(parts[1]);
            const tid = parseInt(parts[2]);
            if (pid === process.pid) {
                if (worker_threads.isMainThread && tid === worker_threads.threadId) {
                    // main thread can't call itself, so this may be a different thread cluster.
                }
                else {
                    return connectIPCObject(clusterObject, parseInt(parts[2]));
                }
            }
        }

        try {
            const clusterPeerPromise = ensureClusterPeer(address, port);
            const clusterPeer = await clusterPeerPromise;
            // may already have this proxy so check first.
            const existing = clusterPeer.remoteWeakProxies[proxyId]?.deref();
            if (existing)
                return existing;
            const newValue = await peerConnectRPCObject(clusterPeer, clusterObject);
            if (!newValue)
                throw new Error('rpc object not found?');
            return newValue;
        }
        catch (e) {
            console.error('failure rpc', clusterObject, e);
            return value;
        }
    }

    const onProxySerialization = (peer: RpcPeer, value: any, sourceKey: string) => {
        const properties = RpcPeer.prepareProxyProperties(value) || {};
        let clusterEntry: ClusterObject = properties.__cluster;

        // ensure globally stable proxyIds.
        // worker threads will embed their pid and tid in the proxy id for cross worker fast path.
        const proxyId = peer.localProxied.get(value)?.id || clusterEntry?.proxyId || `n-${process.pid}-${worker_threads.threadId}-${RpcPeer.generateId()}`;

        // if the cluster entry already exists, check if it belongs to this node.
        // if it belongs to this node, the entry must also be for this peer.
        // relying on the liveness/gc of a different peer may cause race conditions.
        if (clusterEntry) {
            if (isClusterAddress(clusterEntry?.address) && clusterPort === clusterEntry.port && sourceKey !== clusterEntry.sourceKey)
                clusterEntry = undefined;
        }

        if (!clusterEntry) {
            clusterEntry = {
                id: clusterId,
                address: SCRYPTED_CLUSTER_ADDRESS,
                port: clusterPort,
                proxyId,
                sourceKey,
                sha256: null,
            };
            clusterEntry.sha256 = computeClusterObjectHash(clusterEntry, clusterSecret);
            properties.__cluster = clusterEntry;
        }

        return {
            proxyId,
            properties,
        };
    }
    const initializeCluster: InitializeCluster = async (options: {
        clusterId: string;
        clusterSecret: string;
        clusterWorkerId: string;
    }) => {
        if (clusterPort)
            return;

        ({ clusterId, clusterSecret, clusterWorkerId, } = options);

        const clients = new Set<net.Socket>();


        const { server: clusterRpcServer, port } = await clusterListenZero(client => {
            const clusterPeerAddress = client.remoteAddress;
            const clusterPeerPort = client.remotePort;
            const clusterPeerKey = getClusterPeerKey(clusterPeerAddress, clusterPeerPort);
            const clusterPeer = createDuplexRpcPeer(peer.selfName, clusterPeerKey, client, client);
            // set params from the primary peer, needed for get getRemote in cluster mode.
            Object.assign(clusterPeer.params, peer.params);
            // the listening peer sourceKey (client address/port) is used by the OTHER peer (the client)
            // to determine if it is already connected to THIS peer (the server).
            clusterPeer.onProxySerialization = (value) => onProxySerialization(clusterPeer, value, clusterPeerKey);
            clusterPeers.set(clusterPeerKey, Promise.resolve(clusterPeer));
            clusterPeer.params.connectRPCObject = connectClusterObject;
            client.on('close', () => {
                clusterPeers.delete(clusterPeerKey);
                clusterPeer.kill('cluster socket closed');
                clients.delete(client);
            });
            clients.add(client);
        });

        clusterPort = port;

        peer.onProxySerialization = value => onProxySerialization(peer, value, undefined);
        delete peer.params.initializeCluster;

        peer.killedSafe.finally(() => clusterRpcServer.close());
        clusterRpcServer.on('close', () => {
            peer.kill('cluster server closed');
            // close all clusterRpcServer clients
            for (const client of clients) {
                client.destroy();
            }
            clients.clear();
        });
    }

    return {
        initializeCluster,
        get clusterPort() {
            return clusterPort;
        },
        SCRYPTED_CLUSTER_ADDRESS,
        clusterPeers,
        onProxySerialization,
        connectClusterObject,
        ensureClusterPeer,
        mainThreadPort,
        mainThreadBrokerRegister,
        connectRPCObject,
    }
}

export type InitializeCluster = (cluster: { clusterId: string, clusterSecret: string, clusterWorkerId: string, }) => Promise<void>;

export function getScryptedClusterMode(): ['server' | 'client', string, number] {
    const mode = process.env.SCRYPTED_CLUSTER_MODE as 'server' | 'client';

    if (!mode) {
        if (process.env.SCRYPTED_CLUSTER_ADDRESS) {
            console.warn('SCRYPTED_CLUSTER_ADDRESS is set but SCRYPTED_CLUSTER_MODE is not set. This setting will be ignored.');
            delete process.env.SCRYPTED_CLUSTER_ADDRESS;
        }
        if (process.env.SCRPYTED_CLUSTER_SERVER) {
            console.warn('SCRYPTED_CLUSTER_SERVER is set but SCRYPTED_CLUSTER_MODE is not set. This setting will be ignored.');
            delete process.env.SCRPYTED_CLUSTER_SERVER
        }
        if (process.env.SCRYPTED_CLUSTER_SECRET) {
            console.warn('SCRYPTED_CLUSTER_SECRET is set but SCRYPTED_CLUSTER_MODE is not set. This setting will be ignored.');
            delete process.env.SCRYPTED_CLUSTER_SECRET;
        }
        return;
    }

    if (!['server', 'client'].includes(mode))
        throw new Error('SCRYPTED_CLUSTER_MODE must be set to either "server" or "client".');

    if (!process.env.SCRYPTED_CLUSTER_SECRET)
        throw new Error('SCRYPTED_CLUSTER_MODE is set but SCRYPTED_CLUSTER_SECRET is not set.');

    const [server, sport] = process.env.SCRYPTED_CLUSTER_SERVER?.split(':') || [];
    const port = parseInt(sport) || 10556;
    const address = process.env.SCRYPTED_CLUSTER_ADDRESS;

    if (mode === 'client') {
        // server may be a hostname so no IP check is necessary
        if (!server)
            throw new Error('SCRYPTED_CLUSTER_SERVER is not a valid IP address:port.');

        // the address can be determined by checking the socket.localAddress after connection.
        // if (!net.isIP(address))
        //     throw new Error('SCRYPTED_CLUSTER_ADDRESS is not set to a valid IP address.');
    }
    else {
        // the cluster address may come from the server:port combo or address variable but not both.
        if (address && server && server !== address)
            throw new Error('SCRYPTED_CLUSTER_ADDRESS and SCRYPTED_CLUSTER_SERVER must not both be used.');

        let serverAddress = address || server;
        if (!net.isIP(serverAddress)) {
            // due to dhcp changes allowing an interface name for the server address is also valid,
            // resolve using network interfaces.
            const interfaces = os.networkInterfaces();
            const iface = interfaces[serverAddress];
            const ipv4 = iface?.find(i => i.family === 'IPv4');
            if (!ipv4)
                throw new Error('SCRYPTED_CLUSTER_ADDRESS is not set.');
            serverAddress = ipv4.address;
        }
        process.env.SCRYPTED_CLUSTER_ADDRESS = serverAddress;
        delete process.env.SCRYPTED_CLUSTER_SERVER;
    }

    return [mode, server, port];
}

export async function clusterListenZero(callback: (socket: net.Socket) => void) {
    const SCRYPTED_CLUSTER_ADDRESS = process.env.SCRYPTED_CLUSTER_ADDRESS;
    if (!SCRYPTED_CLUSTER_ADDRESS) {
        const server = new net.Server(callback);
        const port = await listenZero(server, '127.0.0.1');
        return {
            server,
            port,
        }
    }

    // need to listen on the cluster address and 127.0.0.1 on the same port.
    let retries = 5;
    while (retries--) {
        const server = new net.Server(callback);
        const port = await listenZero(server, SCRYPTED_CLUSTER_ADDRESS);
        try {
            const localServer = new net.Server(callback);
            localServer.listen(port, '127.0.0.1');
            await once(localServer, 'listening');
            server.on('close', () => localServer.close());
            return {
                server,
                port,
            }
        }
        catch (e) {
            // port may be in use, keep trying.
            server.close();
        }
    }

    throw new Error('failed to bind to cluster address.');
}
